pandora-server-directory/src/networking/manager_client.ts

```diff
@@ -903,3 +934,38 @@ function WithConstantTime<TParams extends unknown[], TReturn extends { /** only
 		return result[1];
 	};
 }
+
+const LoginManager = new class LoginManager {
+	private invalidAttempts: { readonly timestamp: number; }[] = [];
+
+	public loginFailed(): void {
+		const now = Date.now();
+		this.invalidAttempts.push({ timestamp: now });
+		this.invalidAttempts = this.invalidAttempts
+			.slice(-ENV.LOGIN_ATTEMPT_LIMIT)
+			.filter((attempt) => (attempt.timestamp + ENV.LOGIN_ATTEMPT_WINDOW) > now);
+	}
+
+	public async testOptionalCaptcha(data: SecondFactorData | undefined): Promise<null | SecondFactorResponse> {
+		if (!this.isCaptchaRequired()) {
+			return null;
+		}
+		const types: SecondFactorType[] = ['captcha'];
+		if (data == null) {
+			return { result: 'secondFactorRequired', types };
+		}
+		if (data.captcha == null) {
+			this.loginFailed();
+			return { result: 'secondFactorInvalid', types, invalid: [], missing: types };
+		}
+		if (!await TestCaptcha(data.captcha)) {
+			this.loginFailed();
+			return { result: 'secondFactorInvalid', types, invalid: types, missing: [] };
+		}
+		return null;
+	}
+
+	private isCaptchaRequired(): boolean {
+		return this.invalidAttempts.length >= ENV.LOGIN_ATTEMPT_LIMIT && (this.invalidAttempts[0].timestamp + ENV.LOGIN_ATTEMPT_WINDOW) > Date.now();
```

Comment:
This should not trigger if captcha is not configured at all `(!HCAPTCHA_SECRET_KEY || !HCAPTCHA_SITE_KEY)`.